In [2]:
import pymupdf4llm

In [ ]:
md_txt_pdf1_page59 = pymupdf4llm.to_markdown("../data/pdfs/Starting Out with Python, Global Edition, 4th Edition.pdf", pages=[61,62])


md_txt_pdf1_page59

'2.3 Displaying Output with the print Function **59**\n\nthis book are *not* part of the program. We use the line numbers in our discussion to refer\nto parts of the program.)\n\nIt is important to understand that the statements in this program execute in the order that\nthey appear, from the top of the program to the bottom. When you run this program, the first\nstatement will execute, followed by the second statement, and followed by the third statement.\n\n**Strings and String Literals**\n\nPrograms almost always work with data of some type. For example, Program 2-1 uses the\nfollowing three pieces of data:\n\n\'Kate Austen\'\n\n\'123 Full Circle Drive\n\n\'Asheville, NC 28899\'\n\nThese pieces of data are sequences of characters. In programming terms, a sequence of characters that is used as data is called a *string* . When a string appears in the actual code of a program, it is called a *string literal* . In Python code, string literals must be enclosed in quote marks.\nAs mention

2.3 Displaying Output with the print Function **59**\n\nthis book are *not* part of the program. We use the line numbers in our discussion to refer\nto parts of the program.)\n\nIt is important to understand that the statements in this program execute in the order that\nthey appear, from the top of the program to the bottom. When you run this program, the first\nstatement will execute, followed by the second statement, and followed by the third statement.\n\n**Strings and String Literals**\n\nPrograms almost always work with data of some type. For example, Program 2-1 uses the\nfollowing three pieces of data:\n\n\'Kate Austen\'\n\n\'123 Full Circle Drive\n\n\'Asheville, NC 28899\'\n\nThese pieces of data are sequences of characters. In programming terms, a sequence of characters that is used as data is called a *string* . When a string appears in the actual code of a program, it is called a *string literal* . In Python code, string literals must be enclosed in quote marks.\nAs mentioned earlier, the quote marks simply mark where the string data begins and ends.\n\nIn Python, you can enclose string literals in a set of single-quote marks ( \' ) or a set of\n\n"\n\xaddouble-quote marks ( ). The string literals in Program 2-1 are enclosed in single-quote\nmarks, but the program could also be written as shown in Program 2-2.\n\n\n**60** Chapter 2  Input, Processing, and Output\n\nIf you want a string literal to contain either a single-quote or an apostrophe as part of the\nstring, you can enclose the string literal in double-quote marks. For example, Program 2-3\nprints two strings that contain apostrophes.\n\n**Program 2-3  (apostrophe.py)**\n\nLikewise, you can use single-quote marks to enclose a string literal that contains doublequotes as part of the string. Program 2-4 shows an example.\n\n\n\nPython also allows you to enclose string literals in triple quotes (either """ or \'\'\' ). Triplequoted strings can contain both single quotes and double quotes as part of the string. The\nfollowing statement shows an example:\n\nprint("""I\'m reading "Hamlet" tonight.""")\n\nThis statement will print\n\nI\'m reading "Hamlet" tonight.\n\nTriple quotes can also be used to surround multiline strings, something for which single and\ndouble quotes cannot be used. Here is an example:\n\nprint("""One\n\nTwo\n\nThree""")\n\nThis statement will print\n\nOne\n\nTwo\n\nThree\n\n

### Trying out OCR for images with parsing native text (for a particular page)

In [ ]:
import pypdf

def extract_text_from_page(pdf_path, page_number):
    """
    Extract text from a specific page of a PDF file.
    
    Args:
        pdf_path (str): Path to the PDF file
        page_number (int): Page number to extract 
    
    Returns:
        str: Extracted text from the specified page
    """
    reader = pypdf.PdfReader(pdf_path)
    
    # Extract text from the specific page
    page = reader.pages[page_number]
    text = page.extract_text()
    
    return text



# Example usage:
if __name__ == "__main__":
    pdf_path = "../data/pdfs/Starting Out with Python, Global Edition, 4th Edition.pdf"
    
    # Extract from page 0 (first page, 0-indexed)
    output_text = extract_text_from_page(pdf_path, 61)
    print("Text from given page:")
    print(output_text)
    

Text from given page:
 2.3 Displaying Output with the print Function  59
this book are not part of the program. We use the line numbers in our discussion to refer 
to parts of the program.)
Program 2-1  (output.py)
 1  print('Kate Austen')
 2  print('123 Full Circle Drive')
 3  print('Asheville, NC 28899')
Program Output
Kate Austen
123 Full Circle DriveAsheville, NC 28899
It is important to understand that the statements in this program execute in the order that 
they appear, from the top of the program to the bottom. When you run this program, the first statement will execute, followed by the second statement, and followed by the third statement.
Strings and String Literals
Programs almost always work with data of some type. For example, Program 2-1 uses the following three pieces of data:
'Kate Austen'
'123 Full Circle Drive'Asheville, NC 28899'
These pieces of data are sequences of characters. In programming terms, a sequence of char -
acters that is used as data is called a string

### Generate text to markdown

In [ ]:
md_text = pymupdf4llm.to_markdown("../data/pdfs/Starting Out with Python, Global Edition, 4th Edition.pdf")

# print(md_text) 
# Save the markdown text to a file
with open("../data/extracted/markdown/output.md", "w", encoding="utf-8") as f:
    f.write(md_text)

### Generate text compatible with LlamaIndex 

In [7]:
llama_reader = pymupdf4llm.LlamaMarkdownReader()
llama_docs = llama_reader.load_data("../data/pdfs/Starting Out with Python, Global Edition, 4th Edition.pdf")

Successfully imported LlamaIndex


In [8]:
print(f"Number of documents loaded: {len(llama_docs)}")

Number of documents loaded: 748


In [13]:
print(llama_docs[61].text[:1000])  # Print the first 1000 characters of the first document

2.3 Displaying Output with the print Function **59**

this book are *not* part of the program. We use the line numbers in our discussion to refer
to parts of the program.)

It is important to understand that the statements in this program execute in the order that
they appear, from the top of the program to the bottom. When you run this program, the first
statement will execute, followed by the second statement, and followed by the third statement.

**Strings and String Literals**

Programs almost always work with data of some type. For example, Program 2-1 uses the
following three pieces of data:

'Kate Austen'

'123 Full Circle Drive

'Asheville, NC 28899'

These pieces of data are sequences of characters. In programming terms, a sequence of characters that is used as data is called a *string* . When a string appears in the actual code of a program, it is called a *string literal* . In Python code, string literals must be enclosed in quote marks.
As mentioned earlier, the quote marks